### ISSUE TITLE
What IMO should be diagnosed

### ISSUE THREAD

**Issue by jakubjelinek** [marxin/gcc-changelog#1]
\tPR 87488
We should require that component/ part is provided, perhaps even verify the component (shall we allow some alternatives to component names?  E.g. libfortran/ vs. libgfortran/ or tree-optimization/ vs. tree-opt/ is fairly common).

Would be nice if less than two spaces around the name is diagnosed, and more than two spaces too.

Diagnose trailing whitespace.

Complain about tab * not followed by single space.

Incorrect * gcc/cp/parser.c etc. prefixes

\t* not followed by a single space

Space in between ) and :, or no space after :, or two+ spaces after :, or nothing at all after : (such as running mklog and forgetting to fill in), no capital letter after : and space, no dot at the end of line (note, some of these are complicated by : or ) etc. possibly appearing in the description part,
so perhaps let's treat just the \t* lines as starting file entry and \t( lines as starting another function
entry and require e.g. full stop on the line before).

Diagnose entries for files not in the patch or missing entries for files that need to be filled in
(exceptions new files (perhaps just in certain directories), libgo/, gcc/go/frontend/ and a couple of others).

\t*config/arm/arm.md (addsi3_compare_op1): Add 16-bit thumb-2 variants.
- no space after *

\t* gcc/config/riscv/multilib-generator: (canonical_order): Add 'g'.
- both gcc/ prefix incorrect and : (.*):

\t* config/arm/arm_neon.h
\t(vsha1h_u32): Refactor.
- shall we complain about unnecessary line break when it fits just fine?

Perhaps diagnose way too long lines, don't be pedantic about a few chars excess though

\t(aliasing_matching_component_refs_p):  Break out from ...;
- two spaces after :+1: 

I've made some further tweaks e.g. in https://gcc.gnu.org/g:ab20d992c828

**Comment by marxin** [comment]
Thank you Jakub for the ideas.

Let me create a TODO list for the items:

- [ ] `\tPR 87488` - missing component part
- [ ] add list of available components
- [ ] require exactly 2 spaces around commiter name, report different number
- [ ] trailing whitespaces
- [ ] `\t*␣` - require exactly one space
- [ ] incorrect file prefix `* gcc/cp/parser.c` - ChangeLog in `gcc/cp` subdirectory
- [ ] incorrect `: (.*):`
- [ ] too long lines - 120 characters?
- [ ] two spaces after `:`
- [ ] allow `\t\w+` for another author line

Nice to have:
- [ ] allow aliases for component parts, like `tree-opt/12345`
- [ ] complain about too short lines that can eventually fit in a one-line

**Comment by marxin** [comment]
> so perhaps let's treat just the \t* lines as starting file entry and \t( lines as starting another function
> entry and require e.g. full stop on the line before).

One can create something like

```
\t* gcc/config/xyz/aaa.c (function_a_is_long): Start
\tof comment.
\t(second_function)
```

**Comment by marxin** [comment]
> Diagnose entries for files not in the patch or missing entries for files that need to be filled in
> (exceptions new files (perhaps just in certain directories)

I've already implemented this. I would not make any exceptions for new files.

> libgo/, gcc/go/frontend/ and a couple of others).

These rules need to be added.

**Comment by jakubjelinek** [comment]
> I've already implemented this. I would not make any exceptions for new files.

While it isn't problem for myself, I always add : New test. or : New file. etc entries, some maintainers strongly oppose them for gcc/testsuite/, so I thought the acceptable middle-ground would be to require them for changes to existing files but just automatically add corresponding entries for newly added files, which should be discoverable from the commit diff easily.

As for the co-authors, my preference would be only tab + 4 spaces, because otherwise the co-authors aren't aligned below the primary author, so esthetically is not very nice.

**Comment by marxin** [comment]
> While it isn't problem for myself, I always add : New test. or : New file. etc entries, some maintainers strongly oppose them for gcc/testsuite/, so I thought the acceptable middle-ground would be to require them for changes to existing files but just automatically add corresponding entries for newly added files, which should be discoverable from the commit diff easily.

Yes, that would be easy to achieve that. I've added this to my TODO list.

> 
> As for the co-authors, my preference would be only tab + 4 spaces, because otherwise the co-authors aren't aligned below the primary author, so esthetically is not very nice.

Well, we can accept a regex and then do final formatting we want. Let's accept only the script form for now.

**Comment by rearnsha** [comment]
Do you automatically create author lines if missing?  This should be possible.  Also, we'd want "co-authored-by:" to be taken into account in that case.  This shouldn't be too hard, it's essentially the inverse of what was done for the SVN to git conversion.

**Comment by marxin** [comment]
> Do you automatically create author lines if missing?

Yes, similarly for the timestamp (can be detected from `Date: ` header).

> This should be possible. Also, we'd want "co-authored-by:" to be taken into account in that case. > This shouldn't be too hard, it's essentially the inverse of what was done for the SVN to git conversion.

Let me support that too, it would be easy to add.

**Comment by jakubjelinek** [comment]
On Thu, Apr 30, 2020 at 07:24:38AM -0700, marxin wrote:
> > Do you automatically create author lines if missing?
> 
> Yes, similarly for the timestamp (can be detected from `Date: ` header).

date in ChangeLog file should be the commit date rather than authored date
I think.

> > This should be possible. Also, we'd want "co-authored-by:" to be taken into account in that case. > This shouldn't be too hard, it's essentially the inverse of what was done for the SVN to git conversion.
> 
> Let me support that too, it would be easy to add.

Oh, btw, some people put
Co-Authored-By: ...
co-authored-by: ...
and similar comments (whatever-[Bb]y:) at the end of commit log, so it needs to be verified
we don't consider them part of the ChangeLog entry data.

	Jakub

**Comment by marxin** [comment]
> date in ChangeLog file should be the commit date rather than authored date I think.

Yes. We should grab the date from a Git object we will have in a git hook.

**Comment by marxin** [comment]
>  Oh, btw, some people put Co-Authored-By: ... co-authored-by: ... and similar comments (whatever-[Bb]y:) at the end of commit log, so it needs to be verified we don't consider them part of the ChangeLog entry data.

I would support all case insensitive variants of 'Co-Authored-By', which is what Github invented:
https://help.github.com/en/github/committing-changes-to-your-project/creating-a-commit-with-multiple-authors

There are other '*-by' signatures we want to skip:
```
patches/0289-Fix-expectation-and-types-in-acc_get_property-tests.patch:Reviewed-by: Thomas Schwinge  <thomas@codesourcery.com>
patches/0419-Add-acc_device_radeon-to-name_of_acc_device_t-functi.patch:Reviewed-by: Thomas Schwinge <thomas@codesourcery.com>
patches/0713-arc-Don-t-use-if-conversion-when-optimizing-for-size.patch:Signed-off-by: Claudiu Zissulescu <claziss@gmail.com>
patches/0969-PPC64-fix-documentation-for-__builtin_crypto_vpmsum-.patch:Signed-off-by: Carl Love <carll@us.ibm.com>
patches/1082-arc-Update-legitimate-small-data-address.patch:Signed-off-by: Claudiu Zissulescu <claziss@gmail.com>
patches/1160-arc-Update-tumaddsidi4-test.patch:Signed-off-by: Claudiu Zissulescu <claziss@gmail.com>
patches/1327-dwarf-Generate-DIEs-for-external-variables-with-g1-9.patch:Signed-off-by: Alexey Neyman <stilor@att.net>
patches/1937-Remove-fishy-self-assignment-in-omp-low.c-PR94629.patch:Reviewed-by: Thomas Schwinge  <thomas@codesourcery.com>
patches/2008-OpenACC-Avoid-ICE-in-type-cast-async-wait-clauses.patch:Reviewed-by: Thomas Schwinge <thomas@codesourcery.com>
```

**Comment by rearnsha** [comment]
It's inevitable though that folk will pre-write new changelog entries with dates corresponding to when they commit locally, and the final push to upstream can/will never reflect that.  So unless there are git hooks that allow resetting the commit date at the time of push (which might change the SHA value of the commit - I don't know); then there's no real way to recover the push time and we'll have to live with date stamps that are potentially out of order.

**Comment by jakubjelinek** [comment]
Yes, the ChangeLog update script should overwrite the date.  The date actually isn't even required in the commit message, is just as an option how to reliably find the start of the ChangeLog entry part in the commit message.

**Comment by marxin** [comment]
I would recommend people not to write timestamps and we can use either the CommitterDate or `datetime.now`.
